notebooks/writing_a_trace_transform_cpu_offloading.ipynb

```diff
@@ -0,0 +1,868 @@
+    "        raise RuntimeError(\"Received invalid value for `name` - try `thunder` or `thunder_offload`.\")\n",
+    "\n",
+    "    memory_after_model_load = torch.cuda.max_memory_allocated() / 1e9\n",
+    "    print(f\"Peak memory after loading the model : {memory_after_model_load} GB\")\n",
+    "\n",
+    "    a = jmodel(*args, **kwargs)\n",
+    "\n",
+    "    memory_after_forward = torch.cuda.max_memory_allocated() / 1e9\n",
+    "    print(f\"Peak memory after forward the model : {memory_after_forward} GB\")\n",
+    "\n",
+    "    g = torch.rand_like(a)\n",
+    "    actual_grads = torch.autograd.grad(a, model.parameters(), g)\n",
+    "\n",
+    "    memory_after_backward = torch.cuda.max_memory_allocated() / 1e9\n",
+    "    print(f\"Peak memory after backward the model : {memory_after_backward} GB\")\n",
+    "\n",
+    "    del a, g, actual_grads  # Clear data which is not required for benchmark to free some memory.\n",
+    "    gc.collect()\n",
+    "    torch.cuda.empty_cache()\n",
+    "\n",
+    "    measurement = benchmark(jmodel, model, args, kwargs)\n",
+    "    print(f\"Benchmark Timings - mean : {measurement.mean} - median {measurement.median}\")\n",
+    "\n",
+    "    del jmodel, model, cfg, args, kwargs\n",
+    "\n",
+    "    clear_memory()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 14,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "thunder took -\n",
+      "Peak memory after loading the model : 10.311430656 GB\n"
+     ]
+    },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "[W815 16:01:03.874001433 sdp_utils.cpp:455] Warning: 1Torch was not compiled with flash attention. (function operator())\n",
+      "[W815 16:01:03.874014103 sdp_utils.cpp:504] Warning: 1Torch was not compiled with memory efficient attention. (function operator())\n"
+     ]
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Peak memory after forward the model : 39.679975424 GB\n",
+      "Peak memory after backward the model : 47.024185344 GB\n",
+      "Benchmark Timings - mean : 3.9535594288026914 - median 3.9535594288026914\n",
+      "CUDA Memory has been cleared and currently allocated memory is  8.192e-06\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Uncomment this to run the benchmarks.\n",
+    "# print_memory_usage_and_benchmark(\"thunder\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 15,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "thunder_offload took -\n",
+      "Peak memory after loading the model : 10.311430656 GB\n",
+      "Peak memory after forward the model : 14.647877632 GB\n",
+      "Peak memory after backward the model : 32.004153344 GB\n",
+      "Benchmark Timings - mean : 5.257712998200441 - median 5.257712998200441\n",
```

Comment:
Did we want to also add a benchmark for eager as a third point of comparison, or would that be off-topic? Asking because I remember from my experiments that running thunder default vs eager default already results in some peak memory increase. In other words, in order from highest to lowest peak memory usage, it went thunder default > eager > thunder + cpu offload. Just wanted to check if you've seen a similar pattern

Reply:
I thought it would be off-topic as the tutorial (in my mind) is mainly about an example of writing a trace transform.

> In other words, in order from highest to lowest peak memory usage, it went thunder default > eager > thunder + cpu offload. Just wanted to check if you've seen a similar pattern

Yeah, that is the order that I had observed. (Though I noticed that it depends from model to model and it could change to eager > thunder default > thunder + cpu offload).